Cygwin fixes

corngood opened the pull request:
###### Motivation for this change

This was enough to get git building on cygwin.

The changes _should_ all be limited to stdenv.isCygwin.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

corngood committed:
cygwin: rebase fixes

- use fixupOutputsHook to find libs in all outputs
- don't rebase symlinks

corngood committed:
cygwin: force native symlinks for libraries

This will override the existing winsymlinks setting.  nativestrict
will cause ln to fail if it's unable to create a native symlink.
Native symlinks are required for the windows dll loader to find the
libraries.

This script is also used for cross-mingw, but setting CYGWIN
shouldn't cause a problem.

corngood committed:
gzip: remove cygwin hack

The file being patched doesn't exist any more, and gzip now builds
on cygwin without this.

corngood committed:
glib: allow cygwin

corngood committed:
w3m: remove old cygwin patch

corngood committed:
python2: copy stub library to output on cygwin

corngood committed:
libxslt: fix cygwin libtool error

corngood committed:
ncurses: fix cygwin build

corngood committed:
help2man: help wrapper find libintl on cygwin

corngood committed:
bash: fix cygwin build

corngood committed:
perl-LocaleGettext: fix test failure on cygwin by setting $LANG

corngood committed:
procps-ng: allow cygwin

corngood committed:
gpm: allow cygwin

corngood committed:
libiconv: remove old cygwin hacks

corngood committed:
gettext: fix cygwin build

corngood committed:
openssl: fix cygwin build

corngood committed:
zlib: fix cygwin build

corngood commented on pkgs/build-support/setup-hooks/win-dll-link.sh:
```diff
@@ -35,7 +35,7 @@ _linkDLLs() {
         for dllPath2 in "$dllPath" "$(dirname $(readlink "$dllPath" || echo "$dllPath"))"/*.dll; do
             if [ -e ./"$(basename "$dllPath2")" ]; then continue; fi
-            ln -sr "$dllPath2" .
+            CYGWIN+=\ winsymlinks:nativestrict ln -sr "$dllPath2" .
```
This forces cygwin to use real symlinks, which are needed for the windows library loader.  Unfortunately it depends on having elevated permissions in some cases.

corngood commented on lib/systems/parse.nix:
```diff
@@ -71,7 +71,7 @@ rec {
     (x: isExecFormat x.execFormat && all isKernelFamily (attrValues x.families))
   {
-    cygwin  = { execFormat = pe;      families = { inherit /*unix*/ windows-nt; }; };
+    cygwin  = { execFormat = pe;      families = { inherit unix windows-nt; }; };
```
Could also add cygwin support per-package, but it seems to be that cygwin is as unixy as the others here.

corngood commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -7539,15 +7539,17 @@ let self = _self // overrides; _self = with self; {
     };
-  };
+  } // stdenv.lib.optionalAttrs stdenv.isCygwin {
+    LANG="C";
```
Tests fail without this

Mic92 replied:
I think it is safe to set this on all platforms.

corngood replied:
Done

corngood commented on pkgs/stdenv/cygwin/rebase-x86_64.sh:
```diff
@@ -1,10 +1,10 @@
         return
     fi
-    find $out -name "*.dll" | while read DLL; do
+    find "$prefix" -name "*.dll" -type f | while read DLL; do
```
Stop it from rebasing linked libs that have already been rebased.

zimbatm reviewed (approved):
Looking good!

I don't think that cygwin is really supported right now so we might as well merge that in quickly.

Do you have anything specific you want to tackle before merging into master?

corngood commented:
@zimbatm I just wanted to make sure it good a solid review, but I'm not aware of any problems, so I removed WIP.

Unfortunately, there's an non trivial conflict now (the cygwin kernel was removed), so I'll have to resolve that and test with latest master.

zimbatm commented:
Thanks, ping me when you fixed the rebase conflicts

corngood committed:
stdenv: remove leftover 'system' reference

the rest were removed in 1dc6f15de995da2d0351b34c40215336e219cf82

corngood committed:
gdbm: fix cygwin build

corngood committed:
pkgconfig: fix cygwin build

corngood commented:
@Ericson2314 This patch currently adds back the cygwin kernel you recently removed.  The intention was to allow cygwin to inherit 'unix' packages.  Perhaps 'unix' doesn't really make sense as a kernel family...  What do you think?

Ericson2314 commented:
@corngood Thanks for the @-mention and good work! So I originally had a cygwin kernel too, but then, following LLVM's example took it out. For now, perhaps you can modify the definition of isUnix instead?

Also, I've been meaning to deprecate `stdenv.is*`, and use `{build,host,target}Platform.is*` instead, as the latter are more precise. If you could preempt my deprecation by using that, it would be much appreciated.

Ericson2314 commented on pkgs/development/libraries/gdbm/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl }:
+{ stdenv, lib, fetchurl }:
```
For just one use of `lib`, it's more idiomatic to just do `stdenv.lib`.

corngood replied:
There was one other use of `stdenv.lib`, so I shortened it.

Ericson2314 commented on pkgs/development/libraries/libxslt/default.nix:
```diff
@@ -41,6 +41,8 @@ stdenv.mkDerivation rec {
     moveToOutput lib/python2.7 "$py"
   '';
 
+  makeFlags = optionalString stdenv.isCygwin "LDFLAGS=-no-undefined";
```
Could this also be passed to configure?

Ericson2314 commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -22,7 +22,6 @@ let
       ++ [ ./nix-ssl-cert-file.patch ]
       ++ optional (versionOlder version "1.1.0")
           (if stdenv.isDarwin then ./use-etc-ssl-certs-darwin.patch else ./use-etc-ssl-certs.patch)
-      ++ optional stdenv.isCygwin ./1.0.1-cygwin64.patch
```
Did you delete this patch file?

corngood replied:
Done

corngood commented:
@Ericson2314 Rewriting isUnix and removing the unix kernel family seems like a good idea.  I'm thinking it's best to define Unix as a whitelist of:

```
    Darwin       = { kernel = kernels.darwin; };
    Linux        = { kernel = kernels.linux; };
    SunOS        = { kernel = kernels.solaris; };
    FreeBSD      = { kernel = kernels.freebsd; };
    Hurd         = { kernel = kernels.hurd; };
    NetBSD       = { kernel = kernels.netbsd; };
    OpenBSD      = { kernel = kernels.openbsd; };
    Cygwin       = { kernel = kernels.cygwin; abi = abis.cygnus; };
```

Ericson2314 commented:
@corngood yeah feel free to remove the whole notion of "kernel familes" if you like. `matchAttrs` doesn't really provide much notion of "or"---you can either just define `isUnix` as a function predicate, or maybe make a `matchAnyAttrs` that uses list for or?

Ericson2314 commented:
@corngood yeah feel free to remove the whole notion of "kernel familes" if you like. `matchAttrs` doesn't really provide much notion of "or"---you can either just define `isUnix` as a function predicate, or maybe make a `matchAnyAttrs` that uses a list of patterns with the requirement that just one need match?

corngood committed:
cygwin: fix doubleFromSystem for cygwin

corngood committed:
stdenv: remove unix kernel family

System predicate patterns can now be specified as a list of OR'd
attribute sets.

corngood commented:
@Ericson2314 https://github.com/NixOS/nixpkgs/pull/26021/commits/f851fbace46c5f08d0382680135dcd66aaee57e8 is what I came up with for `isUnix`.  If I made all the patterns lists, then `matchAllAttrs` might make sense in `attrsets.nix`, but I left it local for now.

I didn't completely remove kernel families, because there is at least a conceivable use for them, but I can do that if you'd prefer.

Ericson2314 commented on lib/systems/inspect.nix:
```diff
@@ -13,8 +14,8 @@ rec {
 
-    Unix         = { kernel = { families = { inherit (kernelFamilies) unix; }; }; };
     BSD          = { kernel = { families = { inherit (kernelFamilies) bsd; }; }; };
+    Unix         = [ BSD Linux SunOS Hurd Cygwin ];
```
Could you either remove the inner `rec` and do `with patterns;` here?

corngood replied:
I removed the redundant `patterns.`, aside from the string indexed ones.  I'm not aware of a better way to deal with them.  I don't really have strong feelings on this vs. `with patterns;`, but this is slightly more concise.

Ericson2314 replied:
I think it was the string ones that led me to avoid the inner `rec` to begin with, but that was probably an overkill move on my part :).

Ericson2314 commented:
@corngood good decisions on both counts. I just left you one tiny nit---hardly worth mentioning. Other than that, there is the notes on `stdenv.is` vs `hostPlatform.is*` on https://github.com/NixOS/nixpkgs/pull/26021#issuecomment-307406495 . I suppose it's fine to just to deal with that latter, but it would be useful if anyone wants to cross compile for cygwin.

corngood commented:
@Ericson2314 Is https://github.com/NixOS/nixpkgs/pull/26021/commits/974df1f195793b651b54178e60a7e064ea6dd2d9 what you had in mind?  I'm pretty sure those are all correct as `hostPlatform`.

I can squash them into the original changes if you prefer.

Ericson2314 commented:
@corngood a quick skim found one that should actually be `buildPlatform.is*` but the rest I think are good. Thanks for making these changes!

corngood commented:
@Ericson2314 good catch.  I made that one change and squashed each package.

Ericson2314 commented:
CI is failing because travis sucks and the hydra channels are broken for days, sniff. I suppose I could just throw my hands in the error and merge this like other PRs, but I was waiting in hopes the channels would eventually be fixed and then it could be properly tested.

It's really awful that
1. We allow master to be broken
2. We allow merging PRs unrelated to fixing it once it is

Ericson2314 commented:
Restarting travis now that `nixos-unstable` is caught up; it might make sense to rebase too.

Ericson2314 commented:
Hmm, this does look like a mass rebuild. You can use `nix-env -f . -qa --out-path | sort` and diff to track down hash breakages, or just send this PR to `staging` instead.

corngood commented:
While these changes should all still be valid, some of the core packages have been broken again after rebasing on master.  I'll do some more work on it and either fix the non-cygwin hash breakages or retarget to staging.

corngood commented:
I tracked down the cause of libraries being loaded incorrectly from /usr/bin/:

https://github.com/NixOS/nix/pull/1417

So I was able to re-enable checks for a few packages.

I'm still working on the hash changes.

corngood committed:
stdenv: disable -fPIC on cygwin

It's always on, and you get a warning if you specify it

corngood commented:
@Ericson2314 it looks like a lot of the changes come things like `postPatch = optionalString isCygwin "foo"` where postPatch is now "" instead of null.

This seems to be a common pattern for other platforms, and the only way I can think of to fix it is by introducing null, or by using `// optionalAttrs`.  Both of those could cause problems when someone wants to add cases.

Can you think of a better way?  Should I just move this to staging?

Ericson2314 commented:
@corngood I'd do `${if hostPlatform.isCygwin then "postPatch" else null} = ...` and keep targeting `master`. Then after this is merged you can clean up in staging. That way, you get to use your changes faster :). I am doing the same thing for my https://github.com/NixOS/nixpkgs/pull/26007

corngood commented:
@Ericson2314 I can do `postPatch = if isCygwin then "foo" else null` and get the same hash.  That seems saner to me than the if on the left of the =.

Ericson2314 commented:
I like the former because
 - no extraneous `foo = null`s in nix-repl 
 - it's very weirdness is a reminder to fix in staging!

But it really doesn't matter; do whatever you prefer :).

corngood commented:
I've prepared a series of hacks that can be merged without changing any hashes on non-cygwin platforms.  I can put that in as a separate commit, and then revert it and PR the revert to staging.  Do you think that would be preferable to just merging this into stable?  Keep in mind I'll probably have a stream of similar changes as I rebase and get more packages working.

Ericson2314 commented:
Oh you're certainly free to to straight for staging. I *personally* find waiting for staging to be merged annoying, hence the two-step hacks + hash-breaks. But of it's not annoying you, go straight for staging!

corngood commented:
Ok, I'll move it to staging then.  A bunch of stuff broke in the last rebase, so instead of just building `git` and a few other random packages, I'm now trying to build a 'stage 2?' stdenv using `stagesNix` and then I'm going to see if I can build nix itself with that.  That seems a little less arbitrary.

Ericson2314 commented:
Sounds good. Good luck!

corngood commented:
@Ericson2314 rebased on staging.

Building stdenv (especially gcc) is a can of worms.  I'll keep plugging away at it in my spare time, but there's no point in holding this up.

Ericson2314 commented:
@corngood Fair enough. So is this PR just progress then? Or does git still build?

corngood commented:
@Ericson2314 git doesn't currently build due to breakage in ruby, via docbook.

corngood commented:
@Ericson2314 actually looks like `gitMinimal` does build, so it's sufficient for `fetchgit`, which is what I was originally looking for.

Ericson2314 commented:
Sounds great; I'm going to merge! Sorry this took so long, @corngood, but excellent job!

vcunat commented:
@Ericson2314, @corngood: this made all darwin packages evaluate as broken, stopping Hydra builds.  Any ideas?  I bisected the error to 4ac1901d5, but I don't understand the changes enough (yet).

Ericson2314 commented:
Unix should include Darwin, but doesn't.

corngood commented:
Shit, sorry about that.  I prepared a PR to fix isUnix: #27089.  I'm just in the process of reproducing it.

vcunat commented:
Thanks for the quick fixup.  Job counts on Hydra seem OK now.
